### PR TITLE
drivers: spi: spi_pico_pio: Add basic support for SPI via PIO

### PR DESCRIPTION
**Pull request by ThreeEights**
Add fundamental feature support for RP2040 PIO SPI peripherals. This commit implements synchronous transfer with 8-bit MSB format.  Using PIO allows any GPIO pins to be assigned the roles of CS, CLK, MOSI, and MISO.

Optional features not implemented yet:

  - Interrupt based transfer
  - DMA transfer
  - Slave mode
  - Varying word size
  - 3-wire SPI support
  - LSB-first

### CODE REVIEW

**Comment by ThreeEights** [comment]
I have a sample based on the zephyr/samples/sensor/bme280 sample that demonstrates using the PIO SPI driver.  I thought that ought to be a separate commit, but that requires approval of this commit first.  One option is to create a new samples/boards/rpi_pico/spi_pio sample, though the only change is the addition of the rpi_pico.overlay and updating the prj.cfg. 
 The other option is to just update the existing sample with the .overlay.  Please advise on the best steps to include sample code.

**Comment by carlescufi** [comment]
@ThreeEights I am curious, why PIO instead of using the Pico's native SPI IP? Regarding your sample questions, @yonsch should be able to answer that.

**Comment by ThreeEights** [comment]
@carlescufi - There are 3 main reasons for adding PIO SPI support, in addition to the two built-in SPI devices:
1)  Any arrangement of GPIO pins can be used instead of the pre-defined pin assignments.
2)  More than 2 SPI devices can be supported, up to 4.
3)  But the biggest reason is that SPI support through PIO is required to work with the CYW43439 chip on the Raspberry Pi Pico W.  The hard-wired configuration on that board is not supported by the built-in SPI blocks.

**Comment by ThreeEights** [comment]
@iocapa - Would you be willing to join the review for this pull request?
Thanks,
Steve

**Comment by ThreeEights** [comment]
@fabiobaltieri - I saw you triggered a re-run of the Coding Guidelines checks, but that failed due to apt-get not being able to access the Ubuntu repositories.  Is there a way I can restart the job, or do I need assistance from the build team?

Thanks,
Steve

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri - I saw you triggered a re-run of the Coding Guidelines checks, but that failed due to apt-get not being able to access the Ubuntu repositories. Is there a way I can restart the job, or do I need assistance from the build team?

Hey, I retried it again, not sure what level of permissions you need to retry the individual checks but you can always do a dummy `git commit --amend` and push force again to restart the checks. That'll lose the approvals, but if you don't have any that's fine, if you have some I guess ask the reviewers (or #pr-help on Discord). Good call poking me on it, I wasn't tracking this one.

**Comment by fabiobaltieri** [comment]
Hey while we are at it...

> But the biggest reason is that SPI support through PIO is required to work with the CYW43439 chip on the Raspberry Pi Pico W.  The hard-wired configuration on that board is not supported by the built-in SPI blocks.

Does this actually work on the pico W? Thought it was some odd single-wire SPI-ish thing.

**Comment by ThreeEights** [comment]
@fabiobaltieri - Thanks!  That run completed successfully.

You asked:

> Does this actually work on the pico W? Thought it was some odd single-wire SPI-ish thing.

Not yet.  This is a work in progress.  I wanted to get this baseline out there for review while adding the other pieces needed for the Pico W.

**Comment by ThreeEights** [comment]
@soburi - Thanks for the pointer.  Since clock-control didn't appear in any of the SPI drivers I was using for references, I didn't know about it.  (There are so many parts of Zephyr I don't know about!)  I'll try getting the system clock speed through the API and, if that succeeds, I'll push an update.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c]: @soburi - clock-control is not supported on the Raspberry Pi Pico (or Pico W).  Those boards use a fixed clock.  However, looking back at this, it makes more sense to just use the clock frequency from the devicetree, so I've updated those functions to use that instead of calling into the Pico SDK and pushed the change to GitHub.  Thanks for pointing this out.

**Comment by yonsch** [comment]
@ThreeEights Thanks for this PR, I will take a look later. In the meantime:

> But the biggest reason is that SPI support through PIO is required to work with the CYW43439 chip on the Raspberry Pi Pico W. The hard-wired configuration on that board is not supported by the built-in SPI blocks.

The SPI used for the wifi module is a special one wire version of SPI, which I haven't looked at thoroughly enough to say how different it is from normal SPI. It may be similar enough to share code with notmal SPI, but not necessarily so.

In any case, there's another use for a PIO SPI - the PL022 SPI peripheral only supports MSB first communication. The only way to add LSB first SPI is via PIO. So I think any PIO implementation of SPI should support both.

**Comment by ThreeEights** [comment]
@yonsch - Turns out supporting LSB first just involves flipping one bit in a control register.  Do you know of any (inexpensive, readily acquired) peripherals that use SPI in LSB mode?  Changing the driver is simple, but I don't want to push a change without being able to test it.

Thanks,
Steve

**Comment by yonsch** [comment]
> @yonsch - Turns out supporting LSB first just involves flipping one bit in a control register.  Do you know of any (inexpensive, readily acquired) peripherals that use SPI in LSB mode?  Changing the driver is simple, but I don't want to push a change without being able to test it.
> 
> Thanks,
> Steve
> 

I think @petejohanson has a keyboard with LSB first SPI

**Comment by petejohanson** [comment]
Adafruit sells sharp memory display breakouts that use the ls0xx driver and are LSB: https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/display/ls0xx.c#L306

https://www.adafruit.com/product/3502

Can often be found on Mouser/DigiKey too.

**Comment by ThreeEights** [comment]
@yonsch and @petejohanson - Thank you both!  I'll try to get one of those for testing.

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [inline drivers/spi/spi_pico_pio.c:149]: It should better compare with `>=`.
Intuitively, it seems better to reverse the left hand side and right hand side of `||`.

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [inline drivers/spi/spi_pico_pio.c:77]: Use `UINT16_MAX`

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (soburi) [inline dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:20]: This `info` seem to meaningless.

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (soburi) [inline dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:15]: It should better to describe that this array accept only one element.

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (soburi) [inline dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:20]: ditto

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (soburi) [inline dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:25]: ditto

Comment (soburi) [inline dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:25]: ditto^1

**Review by soburi — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:39]: True.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:40]: Ditto.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:41]: Yes, not needed.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:44]: pio and pio_sm are needed in spi_pico_pio_txrx() to identify the associated state machine.  bits can be local to config.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:46]: Gone.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:149]: The maximum and minimum frequency values are allowed, hence the use of "<" and ">".  But you're right about the order of the comparisons.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:173]: I don't believe so.  Both mosi_gpio and miso_gpio are structs within spi_pico_pio_config, with members port and pin.  Those members can be accessed whether or not dev_cfg->mosi_gpio.port is null.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:315]: Comments are gone.  None of interrupt, DMA, or async operation can be configured.

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (ThreeEights) [reply dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:15]: Hmm ... Since the description uses singular nouns, only one element is implied.  This is consistent with definition of other gpios, at least in the .yaml files for SPI, and follows the guidance in [Example phandle-arrays: GPIOs](https://docs.zephyrproject.org/latest/build/dts/phandles.html#example-phandle-arrays-gpios).  However, if you insist, I'll change it.

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (ThreeEights) [reply dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:20]: You're right.

File: dts/bindings/spi/raspberrypi,pico-spi-pio.yaml
Comment (soburi) [reply dts/bindings/spi/raspberrypi,pico-spi-pio.yaml:15]: I got it.
(I'm not a native English speaker, so it's easy to miss this point, sorry.)

**Review by soburi — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:141]: Looks like DIV_ROUND_UP() will do what I need.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:149]: Yep.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:169]: Thanks for pointing that out.  Yes, it makes sense to assert that at compile time, so I moved those checks to SPI_PICO_PIO_INIT().

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:176]: Including this one.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:187]: Yeah, it's a style thing.  Given the frequency with which mosi, miso, cpol, and cpha appear, I think the code is clearer and less cluttered with those variables.  I'd like to keep them, unless you insist they be replaced.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:208]: As I said above, I'd prefer to keep those variables.

**Comment by ThreeEights** [comment]
Updated in response to second round of review comments.

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [reply drivers/spi/spi_pico_pio.c:201]: OK. This is a matter of taste.

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [reply drivers/spi/spi_pico_pio.c:180]: OK. This is a matter of taste.

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [inline drivers/spi/spi_pico_pio.c:73]: Sorry. It my misunderstanding.
The divider allows to set between 1 to 65536, UINT16_MAX is 65536.
Your original code is right.

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [inline drivers/spi/spi_pico_pio.c:42]: It should better to uint32. (Zephyr don't use much `int` excluding return code')

File: drivers/spi/spi_pico_pio.c
Comment (soburi) [inline drivers/spi/spi_pico_pio.c:109]: It should better to uint32. (Zephyr don't use much `int` excluding return code')

**Review by soburi — CHANGES_REQUESTED** [review]
Most parts are no problem.
`UINT16_MAX` is may my wrong point, could you recheck it?

**Comment by soburi** [comment]
@yonsch Could you take look at?

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c]: Okay, I replaced the "int" variable declarations with "uint32_t".  Those should be the only changes in this round.

**Review by yonsch — COMMENTED** [review]
This looks good generally, but I think we should wait for the PIO API to stabilize with #56678 before merging any new PIO based driver.

**Review by fabiobaltieri — COMMENTED** [review]
Hey good stuff, seems like there's going to be more work on rpi pico PIO based drivers, @yonsch do you think that there should be a `tests/drivers/build_all/pio` test to keep all of these built?

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:73]: Ah, yes - I should have checked that header file.  Yes, 65536 is the correct value;  I'll push that change.

File: drivers/spi/spi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_pico_pio.c:104]: Ah, yes - those pesky vestiges of code made obsolete by earlier changes.  I'll clean these up.

**Comment by ThreeEights** [comment]
@yonsch wrote:  "I think we should wait for the PIO API to stabilize with https://github.com/zephyrproject-rtos/zephyr/pull/56678 before merging any new PIO based driver."
This leaves me in a bit of a quandary:  we have additional driver changes backed up waiting for this one to be completed.  Do we have a timeline for resolving https://github.com/zephyrproject-rtos/zephyr/pull/56678?  Should I be working on reviewing that pull request?

**Comment by ThreeEights** [comment]
Yet another push of changes, in response to comments from @soburi and @fabiobaltieri.

**Review by soburi — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

**Comment by soburi** [comment]
@yonsch I think this PR is enough small to rewrite after #56678 merged.
Could you put "Request changed" review if you not want to merge now.

**Comment by fabiobaltieri** [comment]
@yonsch @soburi I've put a DNM tag while you sort that out to get this off the mergeable list, feel free to remove it when ready.

**Comment by fabiobaltieri** [comment]
Actually I have an extra comment, sorry only noticed it now: the naming pattern for rpi pico driver from the MAINTAINERS file is:

https://github.com/zephyrproject-rtos/zephyr/blob/b2af5e87de12c69f9daa8be353a3bcecad9ae364/MAINTAINERS.yml#L2247-L2249

For example the pio UART driver is `drivers/serial/uart_rpi_pico_pio.c`, @ThreeEights could you change this to match that pattern? That would be `drivers/spi/spi_rpi_pico_pio.c`, that way it will get assigned automatically to the correct maintainers.

**Comment by ThreeEights** [comment]
@fabiobaltieri - Hmmm ... are you recommending just changing the name of the drivers/spi/spi_pico_pio.c file (to drivers/spi/spi_rpi_pico_pio.c)?  Or does this also imply changing the compatible as well?  Currently the compatible is "raspberrypi,pico-spi-pio", which is ... um, compatible with the existing "raspberrypi,pico-spi".

**Comment by fabiobaltieri** [comment]
> @fabiobaltieri - Hmmm ... are you recommending just changing the name of the drivers/spi/spi_pico_pio.c file (to drivers/spi/spi_rpi_pico_pio.c)? Or does this also imply changing the compatible as well? Currently the compatible is "raspberrypi,pico-spi-pio", which is ... um, compatible with the existing "raspberrypi,pico-spi".

Just the file, the compatible `raspberrypi,pico-spi-pio` looks fine and coherent with `raspberrypi,pico-uart-pio`. `raspberrypi,pico-spi` is used by the normal hardware SPI (even though the driver ends up being the pl022 one).

**Comment by ThreeEights** [comment]
@soburi @fabiobaltieri @yonsch - The source file drivers/spi/spi_pico_pio.c was renamed as @fabiobaltieri requested.  (And then I messed up the push and had to clean that up.)

**Review by fabiobaltieri — DISMISSED** [review]

File: drivers/spi/spi_rpi_pico_pio.c
Comment (yonsch) [inline drivers/spi/spi_rpi_pico_pio.c:354]: Why is this not handled by pinctrl?

**Review by yonsch — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_rpi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c:46]: Yes - that formatting should match the output of pioasm, with "//" comments replaced with "/* */".

File: drivers/spi/spi_rpi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c:180]: Right - the initializations were incorrectly removed in response to an earlier review comment.

File: drivers/spi/spi_rpi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c:307]: This is a common design pattern for Zephyr SPI drivers that support async operation.  However, that code is nowhere near ready to submit.  I would prefer to leave this as it is, but I'll change it for now if you insist.

File: drivers/spi/spi_rpi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c:354]: Turns out you're right - it IS handled by pinctrl.  That code is redundant;  I deleted it.

**Comment by ThreeEights** [comment]
@soburi @fabiobaltieri @yonsch - Yet another update to incorporate review comments from @yonsch.  Can someone restart the workflows now that the git conflicts have been resolved?

File: drivers/spi/spi_rpi_pico_pio.c
Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c]: Follow-up on this issue:  The comments only align when displayed with 1 tab == 4 spaces.  Since the Zephyr rules require using tabs, not spaces, the alignment of the comments will depend on the tool used to view the source.

Comment (fabiobaltieri) [reply drivers/spi/spi_rpi_pico_pio.c]: The code base is done with 8 spaces, Linux style, fairly common for C.

Comment (fabiobaltieri) [reply drivers/spi/spi_rpi_pico_pio.c]: Mandatory Linux coding-style reference: https://elixir.bootlin.com/linux/latest/source/Documentation/process/coding-style.rst#L21

Comment (ThreeEights) [reply drivers/spi/spi_rpi_pico_pio.c]: @fabiobaltieri - This is why they invented spaces.  :-)

Comment (fabiobaltieri) [reply drivers/spi/spi_rpi_pico_pio.c]: > @fabiobaltieri - This is why they invented spaces. :-)

Thanks, I'll have 8 of them please. :-)

**Comment by ThreeEights** [comment]
@yonsch - Can you please take a look at my changes and let me know if there's anything else required?
@soburi @fabiobaltieri - Can you please review the latest changes?
Thanks!

**Review by fabiobaltieri — DISMISSED** [review]

**Review by soburi — DISMISSED** [review]

**Comment by ThreeEights** [comment]
@yonsch - Could you please review the changes made in response to your review comments?  Thanks!

**Comment by ThreeEights** [comment]
@yonsch - I'd like to complete this pull request, preferably before another merge conflict shows up.  The last response was over 2 weeks ago.  This driver makes minimal use of zephyr/drivers/misc/pio_rpi_pico, so it should be possible to quickly adapt to any changes in [#56678](https://github.com/zephyrproject-rtos/zephyr/pull/56678).  I'm going to ask dev-review to look at this PR and get their recommendation.

File: drivers/spi/CMakeLists.txt
Comment (yonsch) [inline drivers/spi/CMakeLists.txt:44]: Rebase artifact?

**Review by yonsch — CHANGES_REQUESTED** [review]
This seems to need a rebase but otherwise looks good

File: drivers/spi/CMakeLists.txt
Comment (ThreeEights) [reply drivers/spi/CMakeLists.txt:44]: Yes.  Still trying to clean it up ...

**Review by fabiobaltieri — APPROVED** [review]

File: drivers/spi/CMakeLists.txt
Comment (ThreeEights) [reply drivers/spi/CMakeLists.txt:44]: @yonsch - ... Got it!  Can you approve the review?

**Review by yonsch — APPROVED** [review]

**Review by soburi — APPROVED** [review]

**Pull request merged**
